File: server/internal/game/tiles.go
package game

import (
	"math/rand"
	"sort"
	"time"
)

var honorTiles = []string{"dong", "nan", "xi", "bei", "zhong", "fa", "bai"}

func BuildWall() []string {
	wall := make([]string, 0, 136)
	for _, s := range []string{"w", "t", "b"} {
		for n := 1; n <= 9; n++ {
			code := fmtTile(n, s)
			for i := 0; i < 4; i++ {
				wall = append(wall, code)
			}
		}
	}
	for _, h := range honorTiles {
		for i := 0; i < 4; i++ {
			wall = append(wall, h)
		}
	}
	return wall
}

func fmtTile(n int, suit string) string {
	return string(rune('0'+n)) + suit
}

func ShuffleWall(wall []string, seed int64) {
	r := rand.New(rand.NewSource(seed))
	for i := len(wall) - 1; i > 0; i-- {
		j := r.Intn(i + 1)
		wall[i], wall[j] = wall[j], wall[i]
	}
}

func SortTiles(tiles []string) []string {
	out := append([]string(nil), tiles...)
	sort.Slice(out, func(i, j int) bool {
		return tileOrder(out[i]) < tileOrder(out[j])
	})
	return out
}

func tileOrder(code string) int {
	if code == "ghost" {
		return 999
	}
	if len(code) == 2 && code[1] >= 'w' && code[1] <= 'b' {
		si := int(code[1]-'w') * 20
		return si + int(code[0]-'0')
	}
	for i, h := range honorTiles {
		if code == h {
			return 300 + i
		}
	}
	return 500
}

func IsHonor(code string) bool {
	for _, h := range honorTiles {
		if code == h {
			return true
		}
	}
	return false
}

func TileSuit(code string) string {
	if len(code) == 2 && code[1] >= 'w' && code[1] <= 'b' {
		return string(code[1])
	}
	return "honor"
}

func NowSeed() int64 { return time.Now().UnixNano() }

File: js/constants/tiles.js
/** 牌编码 — 与 Go 后端 tile 字符串一致 */

export const SUITS = ['w', 't', 'b'];
export const SUIT_NAMES = { w: '万', t: '筒', b: '条' };

export const WINDS = ['dong', 'nan', 'xi', 'bei'];
export const WIND_NAMES = { dong: '东', nan: '南', xi: '西', bei: '北' };

export const DRAGONS = ['zhong', 'fa', 'bai'];
export const DRAGON_NAMES = { zhong: '中', fa: '发', bai: '白' };

export const HONORS = [...WINDS, ...DRAGONS];

/** 生成完整 136 张牌墙 */
export function buildFullWall() {
  const wall = [];
  for (const s of SUITS) {
    for (let n = 1; n <= 9; n++) {
      const code = `${n}${s}`;
      for (let i = 0; i < 4; i++) wall.push(code);
    }
  }
  for (const h of HONORS) {
    for (let i = 0; i < 4; i++) wall.push(h);
  }
  return wall;
}

export function tileLabel(code) {
  if (!code) return '';
  if (code === 'ghost') return '鬼';
  const m = /^(\d)([wtb])$/.exec(code);
  if (m) return `${m[1]}${SUIT_NAMES[m[2]]}`;
  if (WIND_NAMES[code]) return WIND_NAMES[code];
  if (DRAGON_NAMES[code]) return DRAGON_NAMES[code];
  return code;
}

export function tileSuit(code) {
  const m = /^(\d)([wtb])$/.exec(code);
  return m ? m[2] : 'honor';
}

export function tileRank(code) {
  const m = /^(\d)([wtb])$/.exec(code);
  if (m) return parseInt(m[1], 10);
  if (WINDS.includes(code)) return WINDS.indexOf(code) + 1;
  if (DRAGONS.includes(code)) return DRAGONS.indexOf(code) + 10;
  return 0;
}

export function sortTiles(tiles) {
  const order = (c) => {
    const s = tileSuit(c);
    const si = s === 'w' ? 0 : s === 't' ? 1 : s === 'b' ? 2 : 3;
    return si * 20 + tileRank(c);
  };
  return [...tiles].sort((a, b) => order(a) - order(b));
}

export function isHonor(code) {
  return HONORS.includes(code);
}

export function isTerminalOrHonor(code) {
  if (isHonor(code)) return true;
  const r = tileRank(code);
  return r === 1 || r === 9;
}
